Readme.md
DSDcc
=====

**DSDcc** is a complete rewrite from the original [DSD (Digital Speech Decoder)](https://github.com/szechyjs/dsd) project.

It is rewritten along the following lines:

  - A purely C++ library with a single decoder object at its central point
  - Works by pushing new samples to the decoder at the upper level rather than pulling it from the underlying filesystem (actual file or device) at the lowest level. This facilitates integration with software using it as a true library. This comes especially handy for projects in Qt that cannot afford using pthreads on their own like [gr-dsd](https://github.com/argilo/gr-dsd) does. In fact the main drive for this is to integrate it in a plugin of [SDRangel](https://github.com/f4exb/sdrangel).
  - Works by polling to get possible new MBE or audio samples after new samples have been pushed to the decoder
  - Option to output audio samples as L+R (stereo) samples with L=R as this may facilitate integration
  - A binary that uses this library is provided for integration with other commands that run in a shell. So basically it works only with input / output files possibly being `stdin` / `stdout` to be integrated in a pipe command. There is no direct usage of audio devices nor fancy side reading from or writing to `.wav` or `.mbe` files.
  - `mbelib` usage is optional at compile time. Without `mbelib` only the raw MBE samples can be extracted to be processed outside of DSDcc with the help of a hardware dongle for example thus lifting the possible copyright violations (See next)

These points have been retained from the original:

  - The decoding methods
  - Minimal changes to the options and state structures
  - Input as S16LE samples at a fixed rate of 48kS/s
  - Audio output as S16LE samples at 8kS/s rate directly out of `mbelib` or upsampled to 48kS/s

<h1>Possible copyright issues with mbelib</h1>

While DSDcc is intended to be patent-free, `mbelib` that it uses describes functions that may be covered by one or more U.S. patents owned by DVSI Inc. The source code itself should not be infringing as it merely describes possible methods of implementation. Compiling or using `mbelib` may infringe on patents rights in your jurisdiction and/or require licensing. It is unknown if DVSI will sell licenses for software that uses `mbelib`.

If you are not comfortable with this just do not compile with `mbelib` support and you will still be able to extract the MBE frames and process them outside DSDcc with the help of a hardware dongle for example (e.g. ThumbDV USB dongle). The provided binary `dsdccx` can use such a dongle with [SerialDV](https://github.com/f4exb/serialDV). See the Building section for details.

If you still want `mbelib` support you have to use the `-DUSE_MBELIB=ON` directive on the `cmake` command line and of course you need to have `mbelib` installed in your system.

<h1>Supported formats</h1>

These are a subset of the ones covered by the original DSD project plus other formats. A large part of the code was rewritten more noticeably the symbol timing and the DMR processing improving a lot from the original DSD. For now we have:

  - DMR/MOTOTRBO: ETSI two slot TDMA standard. MOTOTRBO is a popular implementation of this standard.
  - D-Star: developed and promoted by Icom for Amateur Radio customers.
  - dPMR: ETSI narrowband FDMA standard. This is somehow similar to NXDN 2400 Baud rate mode.
  - Yaesu System Fusion (abbreviated YSF): developed and promoted by Yaesu for Amateur Radio customers partly inspired by [gr-ysf](http://hb9uf.github.io/gr-ysf/). Voice full rate with SerialDV is not supported.

Next we may like to add NXDN exploiting similarities with the already implemented dPMR.

<h1>Source code</h1>

<h2>Repository branches</h2>

- master: the production branch
- dev: the development branch

<h1>Building</h1>

As usual with projects based on cmake create a `build` directory at the root of the cloned repository and cd into it.

For `mbelib`support you will need to specify the `-DUSE_MBELIB=ON` directive on the `cmake` command line and you will need to have [mbelib](https://github.com/szechyjs/mbelib) installed in your system. If you use custom installation paths like `/opt/install/mbelib` for example you will need to add the include and library locations to the cmake command line with these directives: `-DLIBMBE_INCLUDE_DIR=/opt/install/mbelib/include -DLIBMBE_LIBRARY=/opt/install/mbelib/lib/libmbe.so`

For DVSI AMBE3000 serial device support (e.g. ThumbDV) in the binary `dsdccx` you will need to install [SerialDV](https://github.com/f4exb/serialDV). Please refer to the `Readme.md` in this package to install SerialDV. If you have SerialDV installed in a custom directory say `/opt/install/serialdv` you will need to add the include and library locations to the cmake command line with these directives: `-DLIBSERIALDV_INCLUDE_DIR=/opt/install/serialdv/include/serialdv -DLIBSERIALDV_LIBRARY=/opt/install/serialdv/lib/libserialdv.so`

So the full cmake command with a custom installation directory and `mbelib`support will look like: `cmake -Wno-dev -DCMAKE_INSTALL_PREFIX=/opt/install/dsdcc -DUSE_MBELIB=ON -DLIBMBE_INCLUDE_DIR=/opt/install/mbelib/include -DLIBMBE_LIBRARY=/opt/install/mbelib/lib/libmbe.so ..`

The full cmake command with a custom installation directory no `mbelib`support and SerialDV support for the binary will look like: `cmake -Wno-dev -DCMAKE_INSTALL_PREFIX=/opt/install/dsdcc -DLIBSERIALDV_INCLUDE_DIR=/opt/install/serialdv/include/serialdv -DLIBSERIALDV_LIBRARY=/opt/install/serialdv/lib/libserialdv.so`

Then:

  - `make` or `make -j8` on a 8 CPU machine
  - `make install`

<h1>Running</h1>

A binary `dsdccx` is produced and gets installed in the `bin` subdirectory of your installation directory. A typical usage is to pipe in the input from a UDP source of discriminator output samples with `socat` and pipe out to `sox` `play` utility to produce some sound:
`socat stdout udp-listen:9999 | /opt/install/dsdcc/bin/dsdccx -i - -fa -o - | play -q -t s16 -r 8k -c 1 -`

You can also run the example files in the `samples` directory. Please refer to the readme in this directory for instructions.

For more details refer to the online help with the `-h` option: `dsdccx -h`

Since version 1.6 dsdccx has the capability of sending regularly the traffic status messages to a file using the `-M` option. See [messagefile.md](messagefile.md) for details.

---
&#9888; (For use with serialDV) Since kernel 4.4.52 the default for FTDI devices (that is in the ftdi_sio kernel module) is not to set it as low latency. This results in the ThumbDV dongle not working anymore because its response is too slow to sustain the normal AMBE packets flow. The solution is to force low latency by changing the variable for your device (ex: /dev/ttyUSB0) as follows:

`echo 1 | sudo tee /sys/bus/usb-serial/devices/ttyUSB0/latency_timer` or `sudo setserial /dev/ttyUSB0 low_latency`

---


<h1>Developpers notes</h1>

<h2>Structure overview</h2>

  - Everything lives in the `DSDcc` namespace
  - The `DSDDecoder` object handles the core functions of synchronization and global orchestration of the decoding. It also hosts the options and state objects. It collaborates with specialized objects that have full access to the decoder public and private areas using the C++ `friend` directive.
  - The options and state objects are the following:
    - The `DSDOpts` object handles the options configuring the behaviour of the decoder
    - The `DSDState` object handles the run time data and data related to the current state of the decoder
  - The `DSDSymbol` object is responsible for symbol and dibit processing. It receives a new sample with its `pushSample()` method. It processes it and when enough samples have been receives it can produce a new symbol that it stores internally.
  - The `DSDMBEDecoder` object is responsible of taking in AMBE frames and producing the final audio output at 8 kS/s. It is a wrapper around the `mbelib` library. It also handles the optional upsampling of audio to 48 kS/s.
  - The objects specialized in the decoding of the various formats are:
    - The `DSDDMR` object is responsible of handling the processing of DMR frames. It uses the service of `DSDMBEDecoder` to produce the final audio output.
    - The `DSDdPMR` object is responsible of handling the processing of dPMR frames. It uses the service of `DSDMBEDecoder` to produce the final audio output.
    - The `DSDDstar` object is responsible of handling the processing of D-Star frames. It uses the service of `DSDMBEDecoder` to produce the final audio output.
    - The `DSDYSF` object is responsible of handling the processing of Yaesu System Fusion frames. It uses the service of `DSDMBEDecoder` to produce the final audio output.    
  - Some utility objects are also defined:
    - The `Descramble` object contains static data and methods mainly used in the decoding of D-Star frames. It is based on Jonathan Naylor G4KLX code.
    - The `DSDFilters` object as the name implies contains methods to perform various forms of DSP filtering.

<h2>Typical integration</h2>

You can look at the source of the `dsdccx` binary to get an idea. Basically it involves the following steps:

 1. Allocate a new `DSDDecoder` object (stack or heap)
 2. Set the options and state object. with some `DSDDecoder` methods.
 3. Prepare the input (open file or stream)
 4. Get a new sample from the stream
 5. Push this sample to the decoder
 6. With `mbelib` support:
   a. Check if any audio output is available and possibly get its pointer and number of samples
   b. Push these samples to the audio device or the output file or stream
 7. With a DVSI AMBE3000 based serial device and SerialDV support:
   a. use DSDcc::DVController helper class with the processDVSerial method
   b. Check if any audio output is available from the helper class and possibly get its pointer and number of samples
   c. Push these samples to the audio device or the output file or stream
 8. Go back to step #5 until a signal is received or some sort of logic brings the loop to an end
 9. Do the cleanup after the loop or in the signal handler (close file, destroy objects...)

Of course this loop can be run in its own thread or remain synchronous with the calling application. Unlike with the original DSD you have the choice.


CHANGELOG
dsdcc (1.9.0-1) unstable; urgency=medium

  * All: tolerate some mismatching symbols in the sync sequence. Implements #36
  * DMR BP: apply only to the 49 bits of data

 -- Edouard Griffiths, F4EXB <f4exb06@gmail.com>  Sat, 25 Jul 2020 10:38:11 +0200

dsdcc (1.8.4-1) unstable; urgency=medium

  * NXDN: take RTCH-C frames into consideration
  * MSVC compatibility

 -- Edouard Griffiths, F4EXB <f4exb06@gmail.com>  Sat, 01 Dec 2018 20:15:15 +0100

dsdcc (1.8.3-1) unstable; urgency=medium

  * NXDN: support full rate (EFR)

 -- Edouard Griffiths, F4EXB <f4exb06@gmail.com>  Thu, 28 Jun 2018 14:15:15 +0200

dsdcc (1.8.2-1) unstable; urgency=medium

  * Fixed compilation errors

 -- Edouard Griffiths, F4EXB <f4exb06@gmail.com>  Sun, 24 Jun 2018 18:15:15 +0200

dsdcc (1.8.1-1) unstable; urgency=medium

  * High pass filter enhancement

 -- Edouard Griffiths, F4EXB <f4exb06@gmail.com>  Sun, 24 Jun 2018 17:15:15 +0200

dsdcc (1.8.0-1) unstable; urgency=medium

  * NXDN implementation

 -- Edouard Griffiths, F4EXB <f4exb06@gmail.com>  Sun, 24 Jun 2018 16:15:15 +0200

dsdcc (1.7.7-1) unstable; urgency=medium

  * Reverted Implement DMR negative

 -- Edouard Griffiths, F4EXB <f4exb06@gmail.com>  Tue, 24 Apr 2018 16:15:15 +0200

dsdcc (1.7.6-1) unstable; urgency=medium

  * Upsampling by 2,3,4,5 with mbelib decoding

 -- Edouard Griffiths, F4EXB <f4exb06@gmail.com>  Mon, 23 Apr 2018 16:15:15 +0200

dsdcc (1.7.5-1) unstable; urgency=medium

  * Implemented DMR negative

 -- Edouard Griffiths, F4EXB <f4exb06@gmail.com>  Sat, 31 Mar 2018 16:15:15 +0200

dsdcc (1.7.4-1) unstable; urgency=medium

  * changed optional audio high pass filter cutoff down to 240 Hz @ -3 dB

 -- Edouard Griffiths, F4EXB <f4exb06@gmail.com>  Tue, 02 Jan 2018 18:15:15 +0100

dsdcc (1.7.3-1) unstable; urgency=medium

  * Added optional audio high pass filter when processing with mbelib

 -- Edouard Griffiths, F4EXB <f4exb06@gmail.com>  Tue, 26 Dec 2017 19:15:15 +0100

dsdcc (1.7.2-1) unstable; urgency=medium

  * Activated compiler warnings and made corresponding fixes

 -- Edouard Griffiths, F4EXB <f4exb06@gmail.com>  Thu, 25 May 2017 23:15:15 +0200

dsdcc (1.7.1-1) unstable; urgency=medium

  * Make PLL to track symbol synchronization optional
  * YSF: fixed FICH code

 -- Edouard Griffiths, F4EXB <f4exb06@gmail.com>  Sun, 07 May 2017 11:15:15 +0200

dsdcc (1.7.0-1) unstable; urgency=medium

  * Implemented PLL to track symbol synchronization
  * Code cleanup with cppcheck and Eclipse possibly fixing occasional segfault

 -- Edouard Griffiths, F4EXB <f4exb06@gmail.com>  Sat, 06 May 2017 11:15:15 +0200

dsdcc (1.6.0-1) unstable; urgency=medium

  * Initial release as a Debian package
  * Code refactoring with some changes in interface from 1.5.5
  * Added options to log formatted messages containing the traffic control information
  * Added options to specify own geographical position

 -- Edouard Griffiths, F4EXB <f4exb06@gmail.com>  Sat, 14 Jan 2017 11:15:15 +0100



ysf.h
///////////////////////////////////////////////////////////////////////////////////
// Copyright (C) 2016 Edouard Griffiths, F4EXB.                                  //
//                                                                               //
// This program is free software; you can redistribute it and/or modify          //
// it under the terms of the GNU General Public License as published by          //
// the Free Software Foundation as version 3 of the License, or                  //
//                                                                               //
// This program is distributed in the hope that it will be useful,               //
// but WITHOUT ANY WARRANTY; without even the implied warranty of                //
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the                  //
// GNU General Public License V3 for more details.                               //
//                                                                               //
// You should have received a copy of the GNU General Public License             //
// along with this program. If not, see <http://www.gnu.org/licenses/>.          //
///////////////////////////////////////////////////////////////////////////////////

#ifndef YSF_H_
#define YSF_H_

#include <string.h>
#include <iostream>
#include <string>

#include "viterbi5.h"
#include "fec.h"
#include "crc.h"
#include "pn.h"
#include "export.h"

namespace DSDcc
{

class DSDDecoder;

class DSDCC_API DSDYSF
{
public:
    typedef enum
    {
        FIHeader,
        FICommunication,
        FITerminator,
        FITest
    } FrameInformation;

    typedef enum
    {
        CMGroupCQ,
        CMRadioID,
        CMReserved,
        CMIndividual
    } CallMode;

    typedef enum
    {
        MRDirectWave,
        MRDownlinUplinkkNotBusy,
        MRDownlinkUplinkBusy,
        MRReserved
    } MessageRouting;

    typedef enum
    {
        DTVoiceData1,
        DTDataFullRate,
        DTVoiceData2,
        DTVoiceFullRate
    } DataType;

    typedef enum
    {
    	FICHNoError,
		FICHErrorGolay,
		FICHErrorCRC
    } FICHError;

#pragma pack(push, 1)
    struct FICH
    {
        FICH() :
            m_reserved(0),
            m_freqDeviation(0),
            m_voipPath(0),
            m_sqlType(0)
        {
            memset(&m_frameInfo[0], 0, 32);
        }

        uint8_t m_frameInfo[2];     //!< 31:30 FI
        uint8_t m_callsignType[2];  //!< 29:28 CS
        uint8_t m_callMode[2];      //!< 27:26 CM
        uint8_t m_blockNumber[2];   //!< 25:24 BN
        uint8_t m_blockTotal[2];    //!< 23:22 BT
        uint8_t m_frameNumber[3];   //!< 21:19 FN
        uint8_t m_frameTotal[3];    //!< 18:17 FT
        uint8_t m_reserved;         //!< 15
        uint8_t m_freqDeviation;    //!< 14    Dev
        uint8_t m_messagePath[3];   //!< 13:11 MR
        uint8_t m_voipPath;         //!< 10    VoIP
        uint8_t m_dataType[2];      //!< 9:8 DT
        uint8_t m_sqlType;          //!< 7     SQL
        uint8_t m_sqlCode[7];       //!< 6:0   SC

        void setBytes(const unsigned char *bytes) {
            memcpy(&m_frameInfo[0], bytes, 32);
        }

        FrameInformation getFrameInformation() const {
            return (FrameInformation) (((m_frameInfo[0]&1)<<1) + (m_frameInfo[1]&1));
        }

        CallMode getCallMode() const {
            return (CallMode) (((m_callMode[0]&1)<<1) + (m_callMode[1]&1));
        }

        int getBlockNumber() const {
            return ((m_blockNumber[0]&1)<<1) + (m_blockNumber[1]&1);
        }

        int getBlockTotal() const {
            return ((m_blockTotal[0]&1)<<1) + (m_blockTotal[1]&1);
        }

        int getFrameNumber() const {
            return ((m_frameNumber[0]&1)<<2) + ((m_frameNumber[1]&1)<<1) + (m_frameNumber[2]&1);
        }

        int getFrameTotal() const {
            return ((m_frameTotal[0]&1)<<2) + ((m_frameTotal[1]&1)<<1) + (m_frameTotal[2]&1);
        }

        bool isNarrowMode() const {
            return (m_freqDeviation & 1) == 1;
        }

        MessageRouting getMessageRouting() const {
            int mrValue = ((m_messagePath[0]&1)<<2) + ((m_messagePath[1]&1)<<1) + (m_messagePath[2]&1);

            if (mrValue < (int) MRReserved) {
                return (MessageRouting) mrValue;
            } else {
                return MRReserved;
            }
        }

        bool isInternetPath() const {
            return (m_voipPath & 1) == 1;
        }

        DataType getDataType() const {
            return (DataType) (((m_dataType[0]&1)<<1) + (m_dataType[1]&1));
        }

        bool isSquelchCodeEnabled() const {
            return (m_sqlType & 1) == 0;
        }

        int getSquelchCode() const {
            return ((m_sqlCode[0]&1)<<6)
                    + ((m_sqlCode[1]&1)<<5)
                    + ((m_sqlCode[2]&1)<<4)
                    + ((m_sqlCode[3]&1)<<3)
                    + ((m_sqlCode[4]&1)<<2)
                    + ((m_sqlCode[5]&1)<<1)
                    + (m_sqlCode[6]&1);
        }

        friend std::ostream &operator<<(std::ostream &output, const FICH& fich)
        {
            output << "FI: " << (int) fich.getFrameInformation()
                    << " CM: " << (int) fich.getCallMode()
                    << " Block#: " << fich.getBlockNumber() << "/" << fich.getBlockTotal()
                    << " Frame#: " << fich.getFrameNumber() << "/" << fich.getFrameTotal()
                    << " Dev: " << (fich.isNarrowMode() ? "narrow" : "wide")
                    << " MR: " << (int) fich.getMessageRouting()
                    << " VoIP: " << (fich.isInternetPath() ? "internet" : "local")
                    << " DT: " << (int) fich.getDataType()
                    << " SQL: " << (fich.isSquelchCodeEnabled() ? "on" : "off")
                    << " SQL code: " << fich.getSquelchCode();
            return output;
        }
    };
#pragma pack(pop)

    explicit DSDYSF(DSDDecoder *dsdDecoder);
    ~DSDYSF();

    void init();
    void process();

    const FICH& getFICH() const { return m_fich; }
    FICHError getFICHError() const { return m_fichError; }
    const char *getDest() const { return m_dest; }
    const char *getSrc() const { return m_src; }
    const char *getDownlink() const { return m_downlink; }
    const char *getUplink() const { return m_uplink; }
    const char *getRem4() const { return m_rem4; }
    bool radioIdMode() const { return m_fich.getCallMode() == CMRadioID; }
    const char *getDestId() const { return m_destId; }
    const char *getSrcId() const { return m_srcId; }

    static const char *ysfChannelTypeText[4];
    static const char *ysfDataTypeText[4];
    static const char *ysfCallModeText[4];

private:

    void processFICH(int symbolIndex, unsigned char dibit);
    void processHeader(int symbolIndex, unsigned char dibit);
    void processVD1(int symbolIndex, unsigned char dibit);
    void processVD2(int symbolIndex, unsigned char dibit);
    void processVD2Voice(int mbeIndex, unsigned char dibit);
    void processVFR(int symbolIndex, unsigned char dibit);
    void processVFRSubHeader(int symbolIndex, unsigned char dibit);
    void processVFRFullIMBE(int symbolIndex, unsigned char dibit);
    void processCSD1(unsigned char *dchBytes);
    void processCSD2(unsigned char *dchBytes);
    void processCSD3_1(unsigned char *dchBytes);
    void processCSD3_2(unsigned char *dchBytes);
    void processAMBE(int mbeIndex, unsigned char dibit);
    void procesVFRFrame(int mbeIndex, unsigned char dibit);
    void storeSymbolDV(unsigned char *mbeFrame, int dibitindex, unsigned char dibit, bool invertDibit = false);

    bool checkCRC16(unsigned char *bits, unsigned long nbBytes, unsigned char *xoredBytes = 0);
    void scrambleVFR(uint8_t out[], uint8_t in[], uint16_t n, uint32_t seed, uint8_t shift);

    DSDDecoder *m_dsdDecoder;
    int m_symbolIndex;                //!< Current symbol index

    unsigned char m_fichRaw[100];     //!< FICH dibits after de-interleave + Viterbi stuff symbols
    unsigned char m_fichGolay[100];   //!< FICH Golay encoded bits + 4 stuff bits + Viterbi stuff bits
    unsigned char m_fichBits[48];     //!< Final FICH + CRC16
    FICH          m_fich;             //!< Validated FICH
    FICHError     m_fichError;        //!< FICH decoding error status

    unsigned char m_dch1Raw[180];     //!< DCH1 dibits after de-interleave
    unsigned char m_dch1Bits[180];    //!< DCH1 bits after de-convolution

    unsigned char m_dch2Raw[180];     //!< DCH2 dibits after de-interleave
    unsigned char m_dch2Bits[180];    //!< DCH2 bits after de-convolution

    unsigned char m_vd2BitsRaw[104];  //!< V/D type 2 VCH+VeCH after de-interleave and de-whitening
    unsigned char m_vd2MBEBits[72];

    unsigned char m_vfrBitsRaw[144];  //!< VFR bits after de-interleave and de-scarambling
    unsigned char m_vfrBits[88];      //!< VFR bits after FEC
    bool m_vfrStart;

    Viterbi5 m_viterbiFICH;
    Golay_24_12 m_golay_24_12;
    CRC m_crc;
    PN_9_5 m_pn;
    unsigned char m_bitWork[48];

    char m_dest[10+1];     //!< Destination callsign from CSD1
    char m_src[10+1];      //!< Source callsign from CSD1
    char m_downlink[10+1]; //!< Downlink callsign from CSD2
    char m_uplink[10+1];   //!< Uplink callsign from CSD2
    char m_rem1[5+1];      //!< Callsign supplementary information #1 from CSD3
    char m_rem2[5+1];      //!< Callsign supplementary information #2 from CSD3
    char m_rem3[5+1];      //!< Callsign supplementary information #3 from CSD3
    char m_rem4[5+1];      //!< Callsign supplementary information #4 from CSD3
    char m_destId[5+1];    //!< Destination radio ID
    char m_srcId[5+1];     //!< Source radio ID

    // AMBE interleave
    const int *w, *x, *y, *z;
    static const int rW[36];
    static const int rX[36];
    static const int rY[36];
    static const int rZ[36];

    static const int m_fichInterleave[100];   //!< FICH symbols interleaving matrix
    static const int m_dchInterleave[180];    //!< DCH symbols interleaving matrix
    static const int m_vd2Interleave[104];    //!< V/D type 2 interleaving matrix
    static const int m_vd2DVSIInterleave[49]; //!< V/D type 2 interleaving matrix for DVSI AMBE3000 chip use
    static const int m_vfrInterleave[144];    //!< VFR interleaving matrix
};

} // namespace DSDcc



#endif /* YSF_H_ */
